airflow/models/baseoperator.py

```diff
@@ -715,31 +715,67 @@ def prepare_template(self):
-                content = getattr(self, attr, None)
-                if content is None:
-                    continue
-                elif isinstance(content, str) and \
-                        any([content.endswith(ext) for ext in self.template_ext]):
-                    env = self.get_template_env()
-                    try:
-                        setattr(self, attr, env.loader.get_source(env, content)[0])
-                    except Exception as e:
-                        self.log.exception(e)
-                elif isinstance(content, list):
-                    env = self.dag.get_template_env()
-                    for i in range(len(content)):
-                        if isinstance(content[i], str) and \
-                                any([content[i].endswith(ext) for ext in self.template_ext]):
-                            try:
-                                content[i] = env.loader.get_source(env, content[i])[0]
-                            except Exception as e:
-                                self.log.exception(e)
+            if not jinja_env:
+                jinja_env = self.get_template_env()
+
+            for attr_name in self.template_fields:
+                attr_value = getattr(self, attr_name)
+                if attr_value is not None:
+                    setattr(self, attr_name, self.load_template_file(attr_value, jinja_env))
+
         self.prepare_template()
 
+    def load_template_file(self, value: Any, jinja_env: jinja2.Environment) -> Any:
+        """
+        Load the file contents for a given value (template path). The value should contain the path to a
+        template file. E.g. "/tmp/whatever.txt" should return the file content (if the file exists). In case
+        of an error, the value is returned as-is, e.g. "/tmp/non_existing_file.txt" is returned as
+        "/tmp/non_existing_file.txt".
+
+        :param value: Path to a file to load. The path may be a collection, e.g. a list storing multiple paths
+            is also parsed.
+        :param jinja_env: Jinja environment to use for loading the file
+        :return: Content of filepath provided in value, or unchanged value if exception occurred
+        """
+
+        if isinstance(value, str) and any(value.endswith(ext) for ext in self.template_ext):
+            try:
+                source, _, _ = jinja_env.loader.get_source(jinja_env, value)
+                return source
+            except Exception as e:
+                self.log.exception(e)
+
+        if isinstance(value, tuple):
+            if type(value) is not tuple:
+                # tuple is namedtuple
+                return value.__class__(*(self.load_template_file(val, jinja_env) for val in value))
+            else:
+                return tuple(self.load_template_file(val, jinja_env) for val in value)
+
+        elif isinstance(value, list):
+            return [self.load_template_file(val, jinja_env) for val in value]
+
+        elif isinstance(value, dict):
```

Comment:
I'm not sure we actually _want_ to do this. Often the templating files is confusing to our users, and making this happen in more places is going to possible change the behaviour of some tasks.

Reply:
Fair point. For the record, currently Airflow supports loading templated files defined in strings and lists of strings. This PR keeps all code related to that functionality in the same place (method), but adds support for other collection types (and cleans up various issues). This PR does not change the behaviour of tasks.

Do you think we should keep templated file loading restricted to only strings and lists, or allow less by dropping list support, or allow more by the changes in this PR?

Reply:
Sorry missed this question.

I think let's not add any _more_ places/data types that render template files.